sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -17,109 +22,279 @@
+    "metaestimator",
+    METAESTIMATORS,
+    ids=[str(row["metaestimator"].__class__.__name__) for row in METAESTIMATORS],
+)
+def test_error_for_other_methods(metaestimator):
+    # This test complements test_warning_for_indicated_methods but checks for
+    # UnsetMetadataPassedError instead of FutureWarning
+    cls = metaestimator["metaestimator"]
+    estimator = metaestimator["estimator"]()
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    X = metaestimator["X"]
+    y = metaestimator["y"]
+    routing_methods = metaestimator["routing_methods"]
+
+    for method_name in routing_methods:
+        if metaestimator.get(method_name, {}).get("warns_on"):
+            # this method is not expected to warn
+            continue
+
+        msg = (
+            "[sample_weight, metadata] are passed but are not explicitly set as "
+            "requested or not for "
+            f"{estimator.__class__.__name__}.{method_name}"
+        )
+        with pytest.raises(UnsetMetadataPassedError, match=re.escape(msg)):
+            method = getattr(instance, method_name)
+            method(X, y, sample_weight=sample_weight, metadata=metadata)
+
+
+@pytest.mark.parametrize(
+    "metaestimator",
+    METAESTIMATORS,
+    ids=[str(row["metaestimator"].__class__.__name__) for row in METAESTIMATORS],
+)
+def test_setting_request_removes_warning_or_error(metaestimator):
+    # When the metadata is explicitly requested, there should be no warning and
+    # no error.
+    def set_request(estimator, method_name):
+        # e.g. call set_fit_request on estimator
+        method = getattr(estimator, f"set_{method_name}_request")
+        method(sample_weight=True, metadata=True)
+
+    cls = metaestimator["metaestimator"]
+    estimator_name = metaestimator["estimator_name"]
+    X = metaestimator["X"]
+    y = metaestimator["y"]
+    routing_methods = metaestimator["routing_methods"]
 
-    metaest = MultiOutput(
-        Estimator()
-        .set_fit_request(sample_weight=True, metadata="alias")
-        .set_partial_fit_request(sample_weight=True, metadata=True)
-    ).fit(X, y_multi, alias=metadata)
-    # if an estimator requests a metadata but it's not passed, no errors are
-    # raised. Therefore here we don't pass `sample_weight` to test nothing is
-    # raised.
-    check_recorded_metadata(
-        metaest.estimators_[0], "fit", sample_weight=None, metadata=metadata
-    )
-
-    metaest.partial_fit(X, y_multi, metadata=metadata)
-    check_recorded_metadata(
-        metaest.estimators_[0], "partial_fit", sample_weight=None, metadata=metadata
-    )
+    for method_name in routing_methods:
+        estimator = metaestimator["estimator"]()
+        set_request(estimator, method_name)
+        instance = cls(**{estimator_name: estimator})
+        # lines below to ensure that there are no warnings
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
```

Comment:
we might need to relax this and catch warnings and check that the relevant warnings are not raised. There might be other warnings like convergence warning, etc., which we don't care about in this context.

Reply:
Okay, I changed the test to record the `FutureWarning`s only, and then check that `"metadata"` is in the message. My reasoning was that we want to avoid raising on other `FutureWarning`s (I guess), so we need to check the message. However, I also didn't want to do an exact check on the whole message. The reason for that is that if someone changes the message but doesn't change this test (which could easily happen), it would pass trivially. I thought checking for `"metadata" is message` would be a good compromise.